### PR TITLE
Pathfinder - approximated paths when original destination unreachable

### LINKED ISSUE #1963: Escaping mobs sometimes do not escape
When hitting a passive mob, it is supposed to escape, sometimes it does not.

I know what the underlying cause is. I'm reporting it to remind myself to fix it and to let others know that the problem is known.

The problem is as follows: When hit, a mob will set a far away point as its destination, if that destination happens to be unreachable (usually happens to be a tree top), the mob will not run away.

### PR DESCRIPTION
**Pull request by SafwatHalaby**
- Mobs now move to a nearby place if no path to target
- Fixed bugs when leaning
- Temporarily fixed Attacking through walls by lowering attackDistance to 1.5 (Someone should check out why the trace at cAggressiveMonster::Tick is failing)
- Fixed mob head shakes

Fixes #1963

### CODE REVIEW

File: src/Mobs/Path.cpp [hunk 1]
```diff
@@ -150,20 +143,26 @@ bool cPath::Step_Internal()
 	cPathCell * CurrentCell = OpenListPop();
 
 	// Path not reachable, open list exauhsted.
-	if (CurrentCell == nullptr)
+	if ((CurrentCell == nullptr) || (m_StepsLeft == 0))
```
Comment (worktycho) [inline src/Mobs/Path.cpp]: `<=` as a precaution?

**Comment by SafwatHalaby** [comment]
Do not merge yet, I may have commited an old code. Checking.

File: src/Mobs/Path.cpp [hunk 2]
```diff
@@ -150,20 +143,26 @@ bool cPath::Step_Internal()
...
 	{
-		FinishCalculation(ePathFinderStatus::PATH_NOT_FOUND);
-		ASSERT(m_Status == ePathFinderStatus::PATH_NOT_FOUND);
-		return true;
+		printf("Steps left %d\n", m_StepsLeft);
```
Comment (worktycho) [inline src/Mobs/Path.cpp]: Use LOGD, then it will only output in debug, and will go into the logfile.

File: src/Mobs/Path.cpp (hunk 1 above)
Comment (SafwatHalaby) [inline src/Mobs/Path.cpp]: I love crashing stuff. Will add an ASSERT as a precaution.

File: src/Mobs/Path.cpp [hunk 3]
```diff
@@ -150,20 +143,26 @@ bool cPath::Step_Internal()
...
+		// Try to find an alternative nearby path, which doesn't exactly match the destination.
+		if (!AttemptToFindAlternative())
+		{
+			printf("Failed to find alternative\n");
```
Comment (worktycho) [inline src/Mobs/Path.cpp]: Again LOGD

**Comment by SafwatHalaby** [comment]
Yes, this is an old code, I have no idea what I messed up /:

File: src/Mobs/Path.cpp (hunk 1 above)
Comment (worktycho) [inline src/Mobs/Path.cpp]: That's an alternative.

**Comment by SafwatHalaby** [comment]
Recovered the right code and rebased.

**Comment by SafwatHalaby** [comment]
The `AttemptToFindAlternative` is undergoing an overhaul, you may disregard the current version.

File: src/Mobs/Path.cpp [hunk 4]
```diff
@@ -149,29 +155,17 @@ bool cPath::Step_Internal()
 {
 	cPathCell * CurrentCell = OpenListPop();
 
-	// Path not reachable, open list exauhsted.
-	if (CurrentCell == nullptr)
+	// Path not reachable.
+	if ((CurrentCell == nullptr))
```
Comment (worktycho) [inline src/Mobs/Path.cpp]: Why the double brackets?

File: src/Mobs/Path.cpp [hunk 5]
```diff
@@ -213,6 +207,76 @@ bool cPath::Step_Internal()
...
+	z = OriginalZ;
+	// Scan a 5x3x5 cube around the original path, starting with a random spot.
+	// TODO: scan a 5x1x5 and If a block in the scan area is air, we scan the solid below instead.
+	for (;;)
+	{
+		for (;;)
+		{
+			for (y = -3; y <= 3; y+= 1)
+			{
+				if ((x == 0) && (y == 0) && (z == 0))
```
Comment (worktycho) [inline src/Mobs/Path.cpp]: Is it problematic to check the original destination? It seems that this only increases the complexity of the code unnecessarily.

File: src/Mobs/Path.cpp [hunk 6]
```diff
@@ -213,6 +207,76 @@ bool cPath::Step_Internal()
...
+	int x, y, z;
+	x = OriginalX;
+	z = OriginalZ;
+	// Scan a 5x3x5 cube around the original path, starting with a random spot.
+	// TODO: scan a 5x1x5 and If a block in the scan area is air, we scan the solid below instead.
+	for (;;)
+	{
+		for (;;)
+		{
+			for (y = -3; y <= 3; y+= 1)
```
Comment (worktycho) [inline src/Mobs/Path.cpp]: Why not declare y here?

File: src/Mobs/Path.cpp [hunk 7]
```diff
@@ -213,6 +207,76 @@ bool cPath::Step_Internal()
...
+	int OriginalX = (rand() % 5) - 2;
+	int OriginalZ = (rand() % 5) - 2;  // A random value between -2 and 2.
+	int x, y, z;
+	x = OriginalX;
+	z = OriginalZ;
+	// Scan a 5x3x5 cube around the original path, starting with a random spot.
+	// TODO: scan a 5x1x5 and If a block in the scan area is air, we scan the solid below instead.
+	for (;;)
+	{
+		for (;;)
```
Comment (worktycho) [inline src/Mobs/Path.cpp]: I think making this `for(int z = OriginalZ; z != OriginalZ; z = (z == 3 ? -2 : z + 1))` would be clearer

Comment (worktycho) [inline src/Mobs/Path.cpp]: And the same for the outer loop.

**Comment by SafwatHalaby** [comment]
Travis failed due to parts of the code I didn't touch...

File: src/Mobs/Path.cpp (hunk 7 above)
Comment (SafwatHalaby) [inline src/Mobs/Path.cpp]: I'm going to completely re-write this function. I'll keep track of the best cell so far and return that instead.

File: src/Mobs/Path.cpp (hunk 5 above)
Comment (SafwatHalaby) [inline src/Mobs/Path.cpp]: I wrote: 

>The AttemptToFindAlternative function is undergoing an overhaul, you may disregard the current version.

**Comment by SafwatHalaby** [comment]
It's as if @worktycho is focusing `AttemptToFindAlternative` on purpose :P

**Comment by SafwatHalaby** [comment]
`printf` will be eventually removed.

**Comment by SafwatHalaby** [comment]
This is not a minor change, I tested it thoroughly but I may have missed things. Please double check me, especially on CHUNK querying stuff.

File: src/Mobs/Monster.cpp [hunk 8]
```diff
@@ -447,21 +510,32 @@ void cMonster::SetPitchAndYawFromDestination()
...
-		SetYaw(Rotation);
+		VectorToEuler(Distance.x, Distance.y, Distance.z, HeadRotation, HeadPitch);
+		if (abs(Rotation-HeadRotation) < 120)  // We're not an owl. If it's more than 90, don't look behind and instead look at the next waypoint.
+		{
+			SetHeadYaw(HeadRotation);
+			SetPitch(-HeadPitch);
+		}
+		else
+		{
+			Distance = m_NextWayPointPosition - GetPosition();
```
Comment (tigerw) [inline src/Mobs/Monster.cpp]: These operations look like what the body rotation calculations above are doing, maybe just reuse BodyDistance if so?

Comment (SafwatHalaby) [inline src/Mobs/Monster.cpp]: Yeah, makes perfect sense.

**Comment by SafwatHalaby** [comment]
Is Coveralls.io stuck?

**Comment by bearbin** [comment]
Yep. I'm going do disable it because it keeps getting stuck and is pretty much useless.

**Comment by SafwatHalaby** [comment]
Squashing to see if Coveralls was disabled.

**Comment by SafwatHalaby** [comment]
Yes, Coveralls is now disabled.

File: src/Mobs/Monster.cpp [hunk 9]
```diff
@@ -269,22 +290,64 @@ bool cMonster::EnsureProperDestination(cChunk & a_Chunk)
...
+				m_FinalDestination.y -= 1;
+			}
 		}
-		m_FinalDestination.y -= 1;
 	}
 
 
+
+
+
```
Comment (madmaxoft) [inline src/Mobs/Monster.cpp]: Don't add so many empty lines within a function body

File: src/Mobs/Monster.h [hunk 10]
```diff
@@ -180,6 +180,13 @@ class cMonster :
 	/** Coordinates for the ultimate, final destination last given to the pathfinder. */
 	Vector3d m_PathFinderDestination;
 
+	/** True if there's no path to target and we're walking to an approximated location. */
+	bool m_NoPathToTarget;
+
+	/** Usually equals m_PathFinderDestination. If there isn't a path to m_PathFinderDestination,
+	Will be the coordinates of a nearby path. */
+	bool m_NoMoreWayPoints;
```
Comment (madmaxoft) [inline src/Mobs/Monster.h]: Coordinates in a `bool`?

Comment (SafwatHalaby) [inline src/Mobs/Monster.h]: Outdated comment for some other idea I had and discarded.

File: src/Mobs/Monster.cpp (hunk 9 above)
Comment (SafwatHalaby) [inline src/Mobs/Monster.cpp]: Fixed.

File: src/Mobs/Monster.h (hunk 10 above)
Comment (SafwatHalaby) [inline src/Mobs/Monster.h]: Fixed.

**Comment by SafwatHalaby** [comment]
@bearbin, Travis is complaining about you:
```
'__APPLE__' is not defined, evaluates to 0 [-Werror,-Wundef]

#elif __APPLE__
```

**Comment by SafwatHalaby** [comment]
I think this was already fixed before my build started, will re-shake Travis and see if it helps.

**Pull request merged**